sdk/include/platform/sunburst/platform-spi.hh

```diff
@@ -0,0 +1,92 @@
+#pragma once
+#include <cdefs.h>
+#include <stdint.h>
+
+/**
+ * A Simple Driver for the Sonata's GPIO.
+ *
+ * Documentation source can be found at:
+ * https://github.com/lowRISC/sonata-system/blob/1a59633d2515d4fe186a07d53e49ff95c18d9bbf/doc/ip/spi.md
+ *
+ * Rendered documentation is served from:
+ * https://lowrisc.org/sonata-system/doc/ip/spi.html
+ */
+struct SonataSpi
+{
+	// Interrupt State Register
+	uint32_t intrState;
+	// Interrupt Enable Register
+	uint32_t intrEnable;
+	// Interrupt Test Register
+	uint32_t intrTest;
+	// Configuration register. Controls how the SPI block transmits
+	uint32_t cfg;
+	// Controls the operation of the SPI block. This register can
+	uint32_t control;
+	// Status information about the SPI block
+	uint32_t status;
+	// When written begins an SPI operation. Writes are ignored when the
+	uint32_t start;
+	// Data from the receive FIFO. When read the data is popped from the
+	uint32_t rxFifo;
+	// Bytes written here are pushed to the transmit FIFO. If the FIFO
+	uint32_t txFifo;
+
+
+	// Initialises the SPI block
+	void
+	init(bool cpol, bool cpha, bool msbFirst, uint16_t halfClkPeriod) volatile
+	{
+		cfg = (cpol ? 1 << 31 : 0) | (cpha ? 1 << 30 : 0) |
+		      (msbFirst ? 1 << 29 : 0) | halfClkPeriod;
+	}
+
+	// Waits for the SPI device to become idle
+	void wait_idle() volatile
+	{
+		// Wait whilst IDLE field in STATUS is low
+		while ((status & 0x40000) == 0) {}
+	}
+
+	// Sends `len` bytes from the given `data` buffer.
+	void tx(const uint8_t data[], uint32_t len) volatile
+	{
+		wait_idle();
+		// Set TX_ENABLE
+		control = 0x4;
+		start   = len;
+
+		uint32_t txAvail = 0;
+		for (int i = 0; i < len; ++i)
+		{
+			if (txAvail == 0)
+			{
+				while (txAvail < 64)
+				{
+					// Read number of bytes in TX FIFO to calculate space
+					// available for more bytes
+					txAvail = 64 - (status & 0xff);
+				}
+			}
+
+			txFifo = data[i];
+			txAvail--;
+		}
+	}
+
+	// Receives `len` bytes and puts them in the `data` buffer.
+	void rx(uint8_t data[], uint32_t len) volatile
+	{
+		wait_idle();
+		// Set RX_ENABLE
+		control = 0x8;
+		start   = len;
+
+		for (int i = 0; i < len; ++i)
+		{
+			// Wait for at least one byte to be available in the RX FIFO
+			while (((status >> 8) & 0xff) == 0) {}
```

Comment:
I don’t really like this spinning without providing an opportunity to yield. Is there not an interrupt that this can block on?

Reply:
There aren't interrupts for the SPI peripheral yet I'm afraid

Reply:
In that case, can we name this method something like `blocking_receive` (or `blocking_read`, for consistency with the UART interface)?  And maybe have a separate method to query whether there's data pending so that people can avoid blocking?

If we can't wait on an interrupt, it would be nice to put a one-tick `thread_sleep` in the wait loop (possibly after a small number of iterations?) so that we're not blocking the entire core until the timer interrupt fires if there's nothing to do.

Reply:
Yes, that sounds a lot nicer